stl/inc/optional

```diff
@@ -53,10 +53,14 @@ protected:
     _THROW(bad_optional_access{});
 }
 
+struct _Nontrivial_Dummy_type {
+    constexpr _Nontrivial_Dummy_type() noexcept {}

```

Comment:
What is the reason you do not use `=default` here? (I assume you have one)

Reply:
Defining the default constructor with `{}` makes it user-provided, therefore non-trivial, therefore the class as a whole is non-trivial. See https://eel.is/c++draft/class.default.ctor#3 and https://eel.is/c++draft/class.prop#2 .

Defining with `= default` makes it user-declared, but not user-provided, see https://eel.is/c++draft/dcl.fct.def.default#5 . (Half of learning the Core Language at an expert level is learning these Words Of Power.)

Reply:
In the specification of *value-initialization* in [[dcl.init]/8](http://eel.is/c++draft/dcl.init#8), [bullet 8.1 states](http://eel.is/c++draft/dcl.init#8.1):

> if `T` is a (possibly cv-qualified) class type ([[class]](http://eel.is/c++draft/class)), then
> * if `T` has either no default constructor ([[class.default.ctor]](http://eel.is/c++draft/class.default.ctor) or a default constructor that is user-provided or deleted, then the object is default-initialized;
> * otherwise, the object is zero-initialized and the semantic constraints for default-initialization are checked, and if `T` has a non-trivial default constructor, the object is default-initialized;

This class has a user-provided default constructor so it matches the first sub-bullet instead of falling into the second which performs zero-initialization.

Reply:
Perhaps this trick deserves a comment. On the one hand, we don't usually explain the Core Language in product code. On the other hand, initialization is notoriously tricky (and I *did* feel the need to explain this in `<charconv>`). Perhaps a short citation to this PR GH-878 would be sufficient.

Reply:
> Perhaps this trick deserves a comment.

I wrote such a comment "This default constructor is user-provided to avoid zero-initialization when objects are value-initialized." after responding here. We don't explain how the core language works, but we do often explain why we use it in particular ways.

Reply:
I need to improve my reading skills, because just 3 day ago I was complaining about code that made the default constructor user provided (via declaring in the header and defaulting in the source file) and blowing up my attempt at a compiler upgrade.

That said I still have the matrix what special member is provided when pinned close at hand